// @angular-mdc/web/bundles/web-theme.umd.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@angular-mdc/web/theme', ['exports', '@angular/core'], factory) :
    (factory((global.ng = global.ng || {}, global.ng.web = global.ng.web || {}, global.ng.web.theme = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MdcContent = /** @class */ (function () {
        function MdcContent() {
            this.isHostClass = true;
        }
        MdcContent.decorators = [
            { type: core.Directive, args: [{
                        selector: 'mdc-content, [mdc-content]'
                    },] },
        ];
        MdcContent.propDecorators = {
            isHostClass: [{ type: core.HostBinding, args: ['class.mdc-content',] }]
        };
        return MdcContent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var MdcMaterialIcons = /** @class */ (function () {
        function MdcMaterialIcons() {
            this.isHostClass = true;
        }
        MdcMaterialIcons.decorators = [
            { type: core.Directive, args: [{
                        selector: '[materialIcons], [material-icon]'
                    },] },
        ];
        MdcMaterialIcons.propDecorators = {
            isHostClass: [{ type: core.HostBinding, args: ['class.material-icons',] }]
        };
        return MdcMaterialIcons;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ THEME_DECLARATIONS = [
        MdcContent,
        MdcMaterialIcons,
    ];
    var MdcThemeModule = /** @class */ (function () {
        function MdcThemeModule() {
        }
        MdcThemeModule.decorators = [
            { type: core.NgModule, args: [{
                        exports: THEME_DECLARATIONS,
                        declarations: THEME_DECLARATIONS,
                    },] },
        ];
        return MdcThemeModule;
    }());

    exports.MdcThemeModule = MdcThemeModule;
    exports.MdcMaterialIcons = MdcMaterialIcons;
    exports.MdcContent = MdcContent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
