// @angular-mdc/web/esm5/theme.es5.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Directive, HostBinding, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var MdcContent = /** @class */ (function () {
    function MdcContent() {
        this.isHostClass = true;
    }
    MdcContent.decorators = [
        { type: Directive, args: [{
                    selector: 'mdc-content, [mdc-content]'
                },] },
    ];
    MdcContent.propDecorators = {
        isHostClass: [{ type: HostBinding, args: ['class.mdc-content',] }]
    };
    return MdcContent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var MdcMaterialIcons = /** @class */ (function () {
    function MdcMaterialIcons() {
        this.isHostClass = true;
    }
    MdcMaterialIcons.decorators = [
        { type: Directive, args: [{
                    selector: '[materialIcons], [material-icon]'
                },] },
    ];
    MdcMaterialIcons.propDecorators = {
        isHostClass: [{ type: HostBinding, args: ['class.material-icons',] }]
    };
    return MdcMaterialIcons;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var /** @type {?} */ THEME_DECLARATIONS = [
    MdcContent,
    MdcMaterialIcons,
];
var MdcThemeModule = /** @class */ (function () {
    function MdcThemeModule() {
    }
    MdcThemeModule.decorators = [
        { type: NgModule, args: [{
                    exports: THEME_DECLARATIONS,
                    declarations: THEME_DECLARATIONS,
                },] },
    ];
    return MdcThemeModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { MdcThemeModule, MdcMaterialIcons, MdcContent };
