// @angular-mdc/web/esm2015/card.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { ChangeDetectionStrategy, Component, Directive, ElementRef, HostBinding, Input, Renderer2, ViewEncapsulation, NgModule } from '@angular/core';
import { toBoolean } from '@angular-mdc/web/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MdcCardMediaContent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
}
MdcCardMediaContent.decorators = [
    { type: Directive, args: [{
                selector: 'mdc-card-media-content, [mdcCardMediaContent]',
                exportAs: 'mdcCardMediaContent'
            },] },
];
/** @nocollapse */
MdcCardMediaContent.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardMediaContent.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__media-content',] }]
};
class MdcCardMedia {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
    /**
     * @return {?}
     */
    get square() { return this._square; }
    /**
     * @param {?} value
     * @return {?}
     */
    set square(value) {
        this.setSquare(value);
    }
    /**
     * @return {?}
     */
    get wide() { return this._wide; }
    /**
     * @param {?} value
     * @return {?}
     */
    set wide(value) {
        this.setWide(value);
    }
    /**
     * @return {?}
     */
    get classSquare() {
        return this.square ? 'mdc-card__media--square' : '';
    }
    /**
     * @return {?}
     */
    get classWide() {
        return this._wide ? 'mdc-card__media--16-9' : '';
    }
    /**
     * @param {?} wide
     * @return {?}
     */
    setWide(wide) {
        this._wide = wide;
    }
    /**
     * @param {?} square
     * @return {?}
     */
    setSquare(square) {
        this._square = square;
    }
}
MdcCardMedia.decorators = [
    { type: Component, args: [{selector: 'mdc-card-media, [mdcCardMedia]',
                template: '<ng-content></ng-content>',
                exportAs: 'mdcCardMedia',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
MdcCardMedia.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardMedia.propDecorators = {
    square: [{ type: Input }],
    wide: [{ type: Input }],
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__media',] }],
    classSquare: [{ type: HostBinding, args: ['class.mdc-card__media--square',] }],
    classWide: [{ type: HostBinding, args: ['class.mdc-card__media--16-9',] }]
};
class MdcCardPrimaryAction {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
}
MdcCardPrimaryAction.decorators = [
    { type: Directive, args: [{
                selector: 'mdc-card-primary-action, [mdcCardPrimaryAction]',
                exportAs: 'mdcCardPrimaryAction'
            },] },
];
/** @nocollapse */
MdcCardPrimaryAction.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardPrimaryAction.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__primary-action',] }]
};
class MdcCardActions {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
    /**
     * @return {?}
     */
    get fullBleed() { return this._fullBleed; }
    /**
     * @param {?} value
     * @return {?}
     */
    set fullBleed(value) {
        this._fullBleed = toBoolean(value);
    }
    /**
     * @return {?}
     */
    get buttons() { return this._buttons; }
    /**
     * @param {?} value
     * @return {?}
     */
    set buttons(value) {
        this._buttons = toBoolean(value);
    }
    /**
     * @return {?}
     */
    get icons() { return this._icons; }
    /**
     * @param {?} value
     * @return {?}
     */
    set icons(value) {
        this._icons = toBoolean(value);
    }
    /**
     * @return {?}
     */
    get classFullBleed() {
        return this.fullBleed ? 'mdc-card__actions--full-bleed' : '';
    }
    /**
     * @return {?}
     */
    get classActionButtons() {
        return this.buttons ? 'mdc-card__action-buttons' : '';
    }
    /**
     * @return {?}
     */
    get classActionIcons() {
        return this.icons ? 'mdc-card__action-icons' : '';
    }
}
MdcCardActions.decorators = [
    { type: Component, args: [{selector: 'mdc-card-actions, [mdcCardActions]',
                template: '<ng-content></ng-content>',
                exportAs: 'mdcCardActions',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
MdcCardActions.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardActions.propDecorators = {
    fullBleed: [{ type: Input }],
    buttons: [{ type: Input }],
    icons: [{ type: Input }],
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__actions',] }],
    classFullBleed: [{ type: HostBinding, args: ['class.mdc-card__actions--full-bleed',] }],
    classActionButtons: [{ type: HostBinding, args: ['class.mdc-card__action-buttons',] }],
    classActionIcons: [{ type: HostBinding, args: ['class.mdc-card__action-icons',] }]
};
class MdcCardActionButtons {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
}
MdcCardActionButtons.decorators = [
    { type: Directive, args: [{
                selector: 'mdc-card-action-buttons, [mdcCardActionButtons]',
                exportAs: 'mdcCardActionButtons'
            },] },
];
/** @nocollapse */
MdcCardActionButtons.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardActionButtons.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__action-buttons',] }]
};
class MdcCardActionIcons {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
}
MdcCardActionIcons.decorators = [
    { type: Directive, args: [{
                selector: 'mdc-card-action-icons, [mdcCardActionIcons]',
                exportAs: 'mdcCardActionIcons'
            },] },
];
/** @nocollapse */
MdcCardActionIcons.ctorParameters = () => [
    { type: ElementRef }
];
MdcCardActionIcons.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__action-icons',] }]
};
class MdcCardAction {
    /**
     * @param {?} _renderer
     * @param {?} elementRef
     */
    constructor(_renderer, elementRef) {
        this._renderer = _renderer;
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
    /**
     * @return {?}
     */
    get action() { return this._action; }
    /**
     * @param {?} action
     * @return {?}
     */
    set action(action) {
        // If the directive is set without a name (updated programatically), then this setter will
        // trigger with an empty string and should not overwrite the programatically set value.
        if (!action) {
            return;
        }
        if (action === 'button') {
            this._renderer.addClass(this.elementRef.nativeElement, 'mdc-card__action--button');
        }
        else if (action === 'icon') {
            this._renderer.addClass(this.elementRef.nativeElement, 'mdc-card__action--icon');
            this._renderer.setAttribute(this.elementRef.nativeElement, 'tabIndex', '0');
            this._renderer.setAttribute(this.elementRef.nativeElement, 'role', 'button');
        }
        this._action = action;
    }
}
MdcCardAction.decorators = [
    { type: Directive, args: [{
                selector: '[mdcCardAction]'
            },] },
];
/** @nocollapse */
MdcCardAction.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
MdcCardAction.propDecorators = {
    action: [{ type: Input, args: ['mdcCardAction',] }],
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card__action',] }]
};
class MdcCard {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.isHostClass = true;
    }
    /**
     * @return {?}
     */
    get outlined() { return this._outlined; }
    /**
     * @param {?} value
     * @return {?}
     */
    set outlined(value) {
        this.setOutlined(value);
    }
    /**
     * @return {?}
     */
    get classOutlined() {
        return this.outlined ? 'mdc-card--outlined' : '';
    }
    /**
     * @param {?} outlined
     * @return {?}
     */
    setOutlined(outlined) {
        this._outlined = outlined;
    }
}
MdcCard.decorators = [
    { type: Component, args: [{selector: 'mdc-card, [mdc-card]',
                template: '<ng-content></ng-content>',
                exportAs: 'mdcCard',
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
MdcCard.ctorParameters = () => [
    { type: ElementRef }
];
MdcCard.propDecorators = {
    outlined: [{ type: Input }],
    isHostClass: [{ type: HostBinding, args: ['class.mdc-card',] }],
    classOutlined: [{ type: HostBinding, args: ['class.mdc-card--outlined',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ CARD_DECLARATIONS = [
    MdcCard,
    MdcCardAction,
    MdcCardActionButtons,
    MdcCardActionIcons,
    MdcCardActions,
    MdcCardMedia,
    MdcCardMediaContent,
    MdcCardPrimaryAction
];
class MdcCardModule {
}
MdcCardModule.decorators = [
    { type: NgModule, args: [{
                exports: CARD_DECLARATIONS,
                declarations: CARD_DECLARATIONS,
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { MdcCardModule, MdcCardMediaContent, MdcCardMedia, MdcCardPrimaryAction, MdcCardActions, MdcCardActionButtons, MdcCardActionIcons, MdcCardAction, MdcCard };
