// @angular-mdc/web/esm2015/theme.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Directive, HostBinding, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MdcContent {
    constructor() {
        this.isHostClass = true;
    }
}
MdcContent.decorators = [
    { type: Directive, args: [{
                selector: 'mdc-content, [mdc-content]'
            },] },
];
MdcContent.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.mdc-content',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MdcMaterialIcons {
    constructor() {
        this.isHostClass = true;
    }
}
MdcMaterialIcons.decorators = [
    { type: Directive, args: [{
                selector: '[materialIcons], [material-icon]'
            },] },
];
MdcMaterialIcons.propDecorators = {
    isHostClass: [{ type: HostBinding, args: ['class.material-icons',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const /** @type {?} */ THEME_DECLARATIONS = [
    MdcContent,
    MdcMaterialIcons,
];
class MdcThemeModule {
}
MdcThemeModule.decorators = [
    { type: NgModule, args: [{
                exports: THEME_DECLARATIONS,
                declarations: THEME_DECLARATIONS,
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { MdcThemeModule, MdcMaterialIcons, MdcContent };
