// @angular-mdc/web/esm2015/web.js
/**
 * @license
 * Copyright (c) 2018 Dominic Carretto
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/trimox/angular-mdc-web/blob/master/LICENSE
 */
import { Version } from '@angular/core';
export { MdcAppBarModule, MdcAppBarRow, MdcAppBarSection, MdcAppBarActionItem, MdcAppBarNavigationIcon, MdcAppBarTitle, MdcAppBarNavSelected, MdcAppBar } from '@angular-mdc/web/app-bar';
export { MdcButtonModule, MdcButton } from '@angular-mdc/web/button';
export { MdcCardModule, MdcCardMediaContent, MdcCardMedia, MdcCardPrimaryAction, MdcCardActions, MdcCardActionButtons, MdcCardActionIcons, MdcCardAction, MdcCard } from '@angular-mdc/web/card';
export { ɵb9, ɵa9, Overlay, OverlayRef, OverlayContainer, OVERLAY_PROVIDERS, OverlayModule } from '@angular-mdc/web/overlay';
export { Portal, ComponentPortal, TemplatePortal, BasePortalOutlet, DomPortalOutlet, CdkPortal, CdkPortalOutlet, PortalModule, PortalInjector } from '@angular-mdc/web/portal';
export { MdcCheckboxModule, MdcCheckboxChange, MDC_CHECKBOX_CONTROL_VALUE_ACCESSOR, MdcCheckbox } from '@angular-mdc/web/checkbox';
export { CHIP_DECLARATIONS, MdcChipsModule, MdcChipSelectionEvent, MdcChipIcon, MdcChipCheckmark, MdcChipText, MdcChip, MdcChipSet } from '@angular-mdc/web/chips';
export { toBoolean, EventRegistry, isSpaceKey, getSupportedInputTypes, UP_ARROW, DOWN_ARROW, RIGHT_ARROW, LEFT_ARROW, PAGE_UP, PAGE_DOWN, HOME, END, ENTER, SPACE, TAB, ESCAPE, BACKSPACE, DELETE, A, Z, ZERO, NINE, COMMA, toNumber, _isNumberValue, isBrowser, MdcPortalService, MdcRouter, MdcRouterModule } from '@angular-mdc/web/common';
export { MdcDialogModule, MdcDialog, MDC_DIALOG_DATA, MdcDialogConfig, MdcDialogContainer, MdcDialogSurface, MdcDialogHeader, MdcDialogHeaderTitle, MdcDialogBody, MdcDialogFooter, MdcDialogBackdrop, MdcDialogButton, MdcDialogRef, MdcDialogComponent } from '@angular-mdc/web/dialog';
export { ɵd14, ɵb14, ɵc14, ɵe14, ɵa14, MdcDrawerModule, FOCUSABLE_ELEMENTS, MdcDrawer } from '@angular-mdc/web/drawer';
export { MdcElevationModule, MdcElevation } from '@angular-mdc/web/elevation';
export { MdcFabModule, MdcFab } from '@angular-mdc/web/fab';
export { MdcFloatingLabelModule, MdcFloatingLabel } from '@angular-mdc/web/floating-label';
export { MdcFormFieldModule, MdcFormField, MdcFormFieldControl } from '@angular-mdc/web/form-field';
export { MdcGridListModule, MdcGridTileTitle, MdcGridTileSupportText, MdcGridTilePrimary, MdcGridTilePrimaryContent, MdcGridTileSecondary, MdcGridTile, MdcGridListTiles, MdcGridList } from '@angular-mdc/web/grid-list';
export { MdcIconModule, MdcIcon } from '@angular-mdc/web/icon';
export { MdcIconButtonModule, MDC_ICON_BUTTON_CONTROL_VALUE_ACCESSOR, MdcIconButtonChange, MdcIconButton } from '@angular-mdc/web/icon-button';
export { MdcIconToggleModule, MD_ICON_TOGGLE_CONTROL_VALUE_ACCESSOR, MdcIconToggle } from '@angular-mdc/web/icon-toggle';
export { MdcImageListModule, MdcImageListItem, MdcImageListImageAspect, MdcImageListImage, MdcImageListSupporting, MdcImageListLabel, MdcImageList } from '@angular-mdc/web/image-list';
export { MdcLineRippleModule, MdcLineRipple } from '@angular-mdc/web/line-ripple';
export { MdcLinearProgressModule, MdcLinearProgress } from '@angular-mdc/web/linear-progress';
export { MdcListModule, MdcListSelectionChange, MdcListItemGraphic, MdcListItemMeta, MdcListItemText, MdcListItemSecondary, MdcListItem, MdcListItemChange, MdcListGroup, MdcListGroupSubheader, MdcListDivider, MdcList } from '@angular-mdc/web/list';
export { MdcMenuModule, MdcMenuChange, MdcMenuAnchor, MdcMenuDivider, MdcMenuItems, MdcMenuItem, MdcMenu } from '@angular-mdc/web/menu';
export { MdcNotchedOutlineModule, MdcNotchedOutlineIdle, MdcNotchedOutline } from '@angular-mdc/web/notched-outline';
export { MdcRadioModule, MDC_RADIO_CONTROL_VALUE_ACCESSOR, MdcRadioChange, MdcRadio } from '@angular-mdc/web/radio';
export { MdcRippleModule, MdcRipple, MdcRippleComponent, MdcRippleDirective } from '@angular-mdc/web/ripple';
export { MdcSelectModule, MDC_SELECT_CONTROL_VALUE_ACCESSOR, MdcSelectChange, MdcSelect } from '@angular-mdc/web/select';
export { MdcShapeModule, MdcShapeContainerCorner, MdcShapeContainer } from '@angular-mdc/web/shape';
export { MdcSliderModule, MDC_SLIDER_CONTROL_VALUE_ACCESSOR, MdcSliderChange, MdcSlider } from '@angular-mdc/web/slider';
export { MdcSnackbarModule, MdcSnackbarConfig, MdcSnackbarRef, MdcSnackbarComponent, MdcSnackbar } from '@angular-mdc/web/snackbar';
export { MdcSwitchModule, MDC_SWITCH_CONTROL_VALUE_ACCESSOR, MdcSwitchChange, MdcSwitch } from '@angular-mdc/web/switch';
export { MdcTabModule, MdcTabIconText, MdcTab, MdcTabChangeEvent, MdcTabBarIndicator, MdcTabBar, MdcTabBarScrollBack, MdcTabBarScrollForward, MdcTabBarScrollFrame, MdcTabBarScroller } from '@angular-mdc/web/tabs';
export { MdcTextFieldModule, MDC_TEXTFIELD_CONTROL_VALUE_ACCESSOR, MdcTextField, MDC_TEXTAREA_CONTROL_VALUE_ACCESSOR, MdcTextarea, MdcTextFieldHelperText } from '@angular-mdc/web/textfield';
export { MdcThemeModule, MdcMaterialIcons, MdcContent } from '@angular-mdc/web/theme';
export { MdcToolbarModule, MdcToolbarIcon, MdcToolbarSection, MdcToolbarTitle, MdcToolbarFixedAdjust, MdcToolbarRow, MdcToolbarMenuIcon, MdcToolbar } from '@angular-mdc/web/toolbar';
export { MdcTypographyModule, MdcTypography, MdcTypographyHeadline1, MdcTypographyHeadline2, MdcTypographyHeadline3, MdcTypographyHeadline4, MdcTypographyHeadline5, MdcTypographyHeadline6, MdcTypographySubtitle1, MdcTypographySubtitle2, MdcTypographyBody2, MdcTypographyBody1, MdcTypographyCaption, MdcTypographyButton, MdcTypographyOverline } from '@angular-mdc/web/typography';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Current version of Angular MDC.
 */
const /** @type {?} */ VERSION = new Version('0.36.0');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { VERSION };
